scenario_player/utils/reclaim.py

```diff
@@ -0,0 +1,520 @@
+    )
+    token_network_deployed_at = BlockNumber(
+        deploy["contracts"][CONTRACT_TOKEN_NETWORK_REGISTRY]["block_number"]
+    )
+
+    # This assumes that reclamation will be done after the scenarios have
+    # executed, so we don't have to synchronize with the whole blockchain, just
+    # up to the last block used by the scenario run. Since we don't know the
+    # exact block, use the current confirmed head.
+    #
+    # This also assumes no RPC calls to pruned data will be performed, so the
+    # target_block is never updated!
+    confirmation_blocks = DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS
+    latest_block = web3.eth.getBlock(BLOCK_ID_LATEST)
+    latest_block_number = latest_block["number"]
+    current_confirmed_head = BlockNumber(latest_block_number - confirmation_blocks)
+
+    # The block range is inclusive, meaning the block represented by
+    # last_fetched_block is skipped, so the block used has to be one before, to
+    # make sure the deployment block is included.
+    start_fetching_at = BlockNumber(token_network_deployed_at - 1)
+
+    all_network_events = _get_all_token_network_events(
+        contract_manager=contract_manager,
+        web3=web3,
+        token_network_address=token_network_address,
+        start_block=start_fetching_at,
+        target_block=current_confirmed_head,
+    )
+
+    tracked_channels: Dict[ChannelID, Dict] = dict()
+
+    # Ignore closed channels and if an address is not under our control. since
+    # The withdraw will only work properly on open channel if performed by both
+    # participants.
+    for event in all_network_events:
+        is_useful_channel_open = (
+            event["event"] == ChannelEvent.OPENED
+            and event["participant1"] in address_to_candidate
+            and event["participant2"] in address_to_candidate
+        )
+        if is_useful_channel_open:
+            tracked_channels[event["channel_identifier"]] = event
+
+        is_useful_channel_close = event["event"] == ChannelEvent.CLOSED and (
+            event["channel_identifier"] in tracked_channels
+        )
+        if is_useful_channel_close:
+            del tracked_channels[event["channel_identifier"]]
+
+    log.debug("Channels found", channels=len(tracked_channels))
+
+    # There must be only one proxy per smart contract and and one JSONRPCClient
+    # per private key, this is necessary for transaction and nonce
+    # synchronization. This loops creates only the necessary instances. For
+    # this to work properly, this function cannot be executed concurrrently.
+    address_to_proxymanager: Dict[Address, ProxyManager] = dict()
+    for channel_open_event in tracked_channels.values():
+        for address in (channel_open_event["participant1"], channel_open_event["participant2"]):
+            if address not in address_to_proxymanager:
+                candidate = address_to_candidate[address]
+                client = JSONRPCClient(web3, candidate.privkey)
```

Comment:
This is the type of code that should be used by the higher layer, to enforce a single instance per address/privatekey.